agent/catalog_endpoint_test.go

```diff
@@ -292,69 +292,60 @@ func TestCatalogNodes_Blocking(t *testing.T) {
 	t.Parallel()
 	a := NewTestAgent(t, t.Name(), "")
 	defer a.Shutdown()
-	testrpc.WaitForTestAgent(t, a.RPC, "dc1")
+	testrpc.WaitForTestAgent(t, a.RPC, "dc1", testrpc.WaitForAntiEntropySync())
 
-	// Register node
+	// Run the query
 	args := &structs.DCSpecificRequest{
 		Datacenter: "dc1",
 	}
-
 	var out structs.IndexedNodes
 	if err := a.RPC("Catalog.ListNodes", *args, &out); err != nil {
 		t.Fatalf("err: %v", err)
 	}
 
-	// t.Fatal must be called from the main go routine
-	// of the test. Because of this we cannot call
-	// t.Fatal from within the go routines and use
-	// an error channel instead.
-	errch := make(chan error, 2)
+	// Async cause a change
+	waitIndex := out.Index
+	start := time.Now()
 	go func() {
-		testrpc.WaitForTestAgent(t, a.RPC, "dc1")
-		start := time.Now()
-
-		// register a service after the blocking call
-		// in order to unblock it.
-		time.AfterFunc(100*time.Millisecond, func() {
-			args := &structs.RegisterRequest{
-				Datacenter: "dc1",
-				Node:       "foo",
-				Address:    "127.0.0.1",
-			}
-			var out struct{}
-			errch <- a.RPC("Catalog.Register", args, &out)
-		})
-
-		// now block
-		req, _ := http.NewRequest("GET", fmt.Sprintf("/v1/catalog/nodes?wait=3s&index=%d", out.Index+1), nil)
-		resp := httptest.NewRecorder()
-		obj, err := a.srv.CatalogNodes(resp, req)
-		if err != nil {
-			errch <- err
+		time.Sleep(100 * time.Millisecond)
+		args := &structs.RegisterRequest{
+			Datacenter: "dc1",
+			Node:       "foo",
+			Address:    "127.0.0.1",
 		}
-
-		// Should block for a while
-		if d := time.Since(start); d < 50*time.Millisecond {
-			errch <- fmt.Errorf("too fast: %v", d)
+		var out struct{}
+		if err := a.RPC("Catalog.Register", args, &out); err != nil {
+			t.Fatalf("err: %v", err)
 		}
+	}()
 
-		if idx := getIndex(t, resp); idx <= out.Index {
-			errch <- fmt.Errorf("bad: %v", idx)
-		}
+	// Re-run the query, if errantly woken up with no change, resume blocking.
+RUN_BLOCKING_QUERY:
+	req, err := http.NewRequest("GET", fmt.Sprintf("/v1/catalog/nodes?wait=3s&index=%d", waitIndex), nil)
```

Comment:
Since the test used to only run the blocking query once with a 3s wait, the time to run the test was capped at ~3s. 

Now the query will loop until it succeeds, or in the worst case until `go test` times out. Should the number of retries be limited?

Reply:
Added a commit that checks that no more time than the original wait duration (3s) has elapsed before looping again. There's no need for padding here on the wait duration because given how the write happens there's not going to be any reason practically that the test is going to need to exhaust the entirety of that duration before returning from the blocking query unless something is broken or way too slow due to mega awful contention.